pkg/cluster/purge_node.go

```diff
@@ -220,19 +220,163 @@ func (c *Controller) removeKubeadmEndpoint(ctx context.Context, name string) (st
+//	  bindPort: 6443
+//
+// apiVersion: kubeadm.k8s.io/v1beta2
+// kind: ClusterStatus
+func marshalClusterStatus(clusterStatus *k8s121ClusterStatus) ([]byte, error) {
+	// deepcopy map
+	apiEndpointsCopy := make(map[string]k8s121APIEndpoint)
+	csApiEndpoints := clusterStatus.apiEndpoints()
+	apiVersion := clusterStatus.apiVersion()
+	kind := clusterStatus.kind()
+
+	for k, v := range csApiEndpoints {
+		apiEndpointsCopy[k] = k8s121APIEndpoint{
+			AdvertiseAddress: v.advertiseAddress(),
+			BindPort:         v.bindPort(),
+		}
+	}
+
+	clusterStatusToWrite := k8s121ClusterStatus{
+		Kind:         kind,
+		APIVersion:   apiVersion,
+		APIEndpoints: apiEndpointsCopy,
+	}
+
+	clusterStatusYaml, err := yaml.Marshal(&clusterStatusToWrite)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal kubeadm-config ClusterStatus")
+	}
+	return clusterStatusYaml, nil
+}
+
+// k8s121ClusterStatus represents a kubeadm.k8s.io/v1beta2 ClusterStatus config
+// It is used to decode the following YAML documents:
+// ---
+// apiendpoints:
+//
+//	rafael-kurl-ecko-purge-master:
+//	  advertiseaddress: 10.128.0.126
+//	  bindport: 6443
+//	rafael-kurl-ecko-purge-master-2:
+//	  advertiseaddress: 10.128.0.63
+//	  bindport: 6443
+//
+// typemeta:
+//
+//	apiversion: kubeadm.k8s.io/v1beta2
+//	kind: ClusterStatus
+//
+// ---
+// apiEndpoints:
+//
+//	rafael-kurl-ecko-purge-master:
+//	  advertiseAddress: 10.128.0.126
+//	  bindPort: 6443
+//	rafael-kurl-ecko-purge-master-2:
+//	  advertiseAddress: 10.128.0.63
+//	  bindPort: 6443
+//
+// apiVersion: kubeadm.k8s.io/v1beta2
+// kind: ClusterStatus
 type k8s121ClusterStatus struct {
-	metav1.TypeMeta
-	APIEndpoints map[string]k8s121APIEndpoint
+	metav1.TypeMeta `yaml:",omitempty"`
+	Kind            string
+	APIVersion      string                       `yaml:"apiVersion"`
+	APIEndpoints    map[string]k8s121APIEndpoint `yaml:"apiEndpoints"`
+	APIEndpointsLC  map[string]k8s121APIEndpoint `yaml:"apiendpoints,omitempty"`
+}
+
+func (k k8s121ClusterStatus) kind() string {
+	kind := k.Kind
+	if k.TypeMeta.Kind != "" {
+		kind = k.TypeMeta.Kind
+	}
+	return kind
+}
+
+func (k k8s121ClusterStatus) apiVersion() string {
+	apiVersion := k.APIVersion
+	if k.TypeMeta.APIVersion != "" {
+		apiVersion = k.TypeMeta.APIVersion
+	}
+	return apiVersion
+}
+
+func (k k8s121ClusterStatus) apiEndpoints() map[string]k8s121APIEndpoint {
+	endpoints := k.APIEndpoints
+	if k.APIEndpointsLC != nil {
+		endpoints = k.APIEndpointsLC
+	}
+	return endpoints
 }
 
 type k8s121APIEndpoint struct {
-	AdvertiseAddress string
-	BindPort         int32
+	AdvertiseAddress   string `yaml:"advertiseAddress"`
+	AdvertiseAddressLC string `yaml:"advertiseaddress,omitempty"`
+	BindPort           int32  `yaml:"bindPort"`
+	BindPortLC         int32  `yaml:"bindport,omitempty"`
+}
+
+func (k k8s121APIEndpoint) advertiseAddress() string {
+	advertiseAddress := k.AdvertiseAddress
+	if k.AdvertiseAddressLC != "" {
+		advertiseAddress = k.AdvertiseAddressLC
```

Comment:
im curious why choose lowercase over upper if both are not empty? or is this a use case we do not have to consider?